### DISCUSSION TITLE
How do I implement redis for storedSessions?

### DISCUSSION THREAD

**Discussion by mkastner** [keystonejs/keystone#8194] [Questions] [answered]
I don't understand this:

[https://keystonejs.com/docs/config/session#stateless-vs-stored-sessions](https://keystonejs.com/docs/config/session#stateless-vs-stored-sessions)

As far as I understand, I need to do something like this:

import { config } from '@keystone-6/core';
import { statelessSessions, storedSessions } from '@keystone-6/core/session';

```
export default config({
  // Stored
  session: storedSessions({ store: { /* ... */ }, /* ... */ }),
  /* ... */
});
```

and if I want to use redis session store, I need to do something like this, I guess:

```
import { redisSessionStore } from "@keystone-6/session-store-redis";
import { createClient } from "redis";

const redisHost = process.env.REDIS_HOST || 'localhost';

const session = storedSessions({ 
  store: redisSessionStore({ //error in here
    client: createClient({ url: `redis://${redisHost}/` }),
  })
});

export default config({
  // Stored
  session,
});
```

Now, the documentation says:

[https://keystonejs.com/docs/config/session#session-stores](https://keystonejs.com/docs/config/session#session-stores)

> ... This store option must be either a SessionStore object, or a function ({ maxAge }) => { ... } which returns a SessionStore object.

> A SessionStore object has the following interface:

```
const store = {
  set: async (sessionId, data) => { /* ... */ },
  get: async (sessionId) => {
    /* ... */
    return data;
  },
  delete: async (sessionId) => { /* ... */ },
};
```

I am wondering about three things here:

1. Where does maxAge come from? I haven't set it yet.
2. Maybe I am supposed to set maxAge in some options, but in which?
```
const session = storedSessions({ 
  // set maxAge here?
  store: redisSessionStore({ //error in here
    // set maxAge here?
  })
});
```
3. Regarding the SessionStore interface:
There's not much information about the implementation and options of session-store-redis:
https://www.npmjs.com/package/@keystone-6/session-store-redis
Does redisSessionStore come with set, get, delete being already implement or do I need to implement these functions (somehow) on redisSessionStore?

**Comment by flexdinesh** [comment, marked as answer]
Thanks for the feedback @mkastner. We will look at improving our session docs to add more clarity. We have two examples to demonstrate custom session implementation. One of that might give you the clarity you are looking for.

1. [Feature Example - Invalidate Session Token](https://github.com/keystonejs/keystone/tree/main/examples/custom-session-validation)
2. [Feature Example - Redis Session Store](https://github.com/keystonejs/keystone/tree/main/examples/custom-session-redis)

**Comment by mkastner** [comment]
@flexdinesh Thanks for your reply. My bad, I had used session-store-redis about which I've meanwhile found out that it's been deprecated.
